### PR TITLE
cmd/identity: allow using redis for RevocationDB

### PR DESCRIPTION
**Pull request by jenlij**
What: Makes sure revocation db has redis support. Sets up storj-sim run redis-server for live accounting and revocation db.

Why: We need a central db for revocation when we split up the satellites.

Tests: Added TestRevocationDB_List to test this method works for both redis and bolt

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: storage/redis/client.go [hunk 1]
```diff
@@ -128,9 +127,14 @@ func (client *Client) GetAll(ctx context.Context, keys storage.Keys) (_ storage.
 	}
 
 	results, err := client.db.MGet(keyStrings...).Result()
+	if len(results) == 0 && err.Error() == "ERR wrong number of arguments for 'mget' command" {
+		fmt.Println("No Revocations Present")
```
Comment (egonelbre) [inline storage/redis/client.go]: using `fmt` here

File: storage/redis/client.go [hunk 2]
```diff
@@ -128,9 +127,14 @@ func (client *Client) GetAll(ctx context.Context, keys storage.Keys) (_ storage.
 	}
 
 	results, err := client.db.MGet(keyStrings...).Result()
+	if len(results) == 0 && err.Error() == "ERR wrong number of arguments for 'mget' command" {
```
Comment (egonelbre) [inline storage/redis/client.go]: why would it return "wrong number of arguments"?

File: cmd/identity/revocations.go [hunk 3]
```diff
@@ -39,10 +38,6 @@ func init() {
 
 func cmdRevocations(cmd *cobra.Command, args []string) error {
 	ctx, _ := process.Ctx(cmd)
-	if len(args) > 0 {
-		revCfg.RevocationDBURL = "bolt://" + filepath.Join(configDir, args[0], "revocations.db")
```
Comment (egonelbre) [inline cmd/identity/revocations.go]: Can this be removed? @bryanchriswhite

Comment (bryanchriswhite) [reply cmd/identity/revocations.go]: I'm pretty sure we need this for this command to work. What's the thinking here?

File: storage/redis/client.go (hunk 1 above)
Comment (bryanchriswhite) [reply storage/redis/client.go]: This redis client isn't specific to revocationDB. Unless I'm misunderstanding something, I don't think anything revocationDB-specific should go here.

Comment (jenlij) [reply storage/redis/client.go]: oh good point

File: cmd/identity/revocations.go (hunk 3 above)
Comment (jenlij) [reply cmd/identity/revocations.go]: I want to be able to pass in either a bolt or redis url. With the current setup, I can't pass in a redis url. I can modify it so it can use the above as a default for bolt if that's preferred.

File: storage/redis/client.go (hunk 1 above)
Comment (jenlij) [reply storage/redis/client.go]: I added this so when I run `identity revocations --revocation-dburl="redis://127.0.0.1:6379?db=0" I don't get that mget error message that occurs if there are 0 items the db. Am I missing something?

File: cmd/identity/revocations.go (hunk 3 above)
Comment (bryanchriswhite) [reply cmd/identity/revocations.go]: Ah, I see. So the idea here was to allow a use to pass a single argument that is the config subdirectory on which the bolt db lives. If you're using redis you wouldn't pass an argument, instead you would just pass the redis url in the flag. I think this can stay as is.

File: storage/redis/client.go (hunk 1 above)
Comment (bryanchriswhite) [reply storage/redis/client.go]: I was not aware of that error. Perhaps it would be better to check for that error in the revocation db package. Also, if the redis package exports that error maybe we can check for it not directly than comparing the error string.

File: cmd/identity/revocations.go (hunk 3 above)
Comment (jenlij) [reply cmd/identity/revocations.go]: oh you're right, adding that check back!

File: pkg/revocation/db_test.go [hunk 4]
```diff
@@ -151,3 +151,38 @@ func TestRevocationDB_Put_error(t *testing.T) {
...
 }
+
+func TestRevocationDB_List(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	testrevocation.RunDBs(t, func(t *testing.T, revDB extensions.RevocationDB, db storage.KeyValueStore) {
+		keys, chain, err := testpeertls.NewCertChain(2, storj.LatestIDVersion().Number)
+		require.NoError(t, err)
+		keys2, chain2, err := testpeertls.NewCertChain(2, storj.LatestIDVersion().Number)
```
Comment (bryanchriswhite) [inline pkg/revocation/db_test.go]: I don't think you need a second chain. You can just revoke the leaf of the one chain multiple times.

File: pkg/revocation/db_test.go [hunk 5]
```diff
@@ -151,3 +151,38 @@ func TestRevocationDB_Put_error(t *testing.T) {
...
+		// list 1,2 revocations
+		firstRevocation, err := extensions.NewRevocationExt(keys[peertls.CAIndex], chain[peertls.LeafIndex])
+		require.NoError(t, err)
+
+		err = revDB.Put(ctx, chain, firstRevocation)
+		require.NoError(t, err)
+		revs, err = revDB.List(ctx)
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(revs))
+
```
Comment (bryanchriswhite) [inline pkg/revocation/db_test.go]: It probably wouldn't be the worst thing to compare the actual revocations, just to ensure nothing weird is going on with `list`. To do this you could marshal the `*Revocation` that's returned by `revDB.List` and compare that agains the `Value` of the extension that's returned by `extensions.NewRevocationExt`; like what `TestRevocationDB_Get` is doing:
```
			revBytes, err := rev.Marshal()
			assert.NoError(t, err)
			assert.True(t, bytes.Equal(ext.Value, revBytes))
```

File: pkg/revocation/db_test.go (hunk 4 above)
Comment (jenlij) [reply pkg/revocation/db_test.go]: I tried using the same chain but it only produces 1 revocation. Is this correct or am I missing something?
`firstRevocation, err := extensions.NewRevocationExt(keys[peertls.CAIndex], chain[peertls.LeafIndex])`
`secondRevocation, err := extensions.NewRevocationExt(keys[peertls.CAIndex], chain[peertls.LeafIndex])`

Comment (bryanchriswhite) [reply pkg/revocation/db_test.go]: That actually makes perfect sense, my mistake. Here we're just creating a revocation extension, not actually modifying the chain by doing a full, proper, revocation. Also, I'm now realizing that revDB will only store the most recent revocation for the same nodeID. So even if we were modifying the chain you would still only have one revocation in the DB. what you did before was correct, sorry for the confusion.

Comment (jenlij) [reply pkg/revocation/db_test.go]: ok no worries, good to know, thanks!

File: cmd/storj-sim/network.go [hunk 6]
```diff
@@ -264,6 +264,11 @@ func newNetwork(flags *Flags) (*Processes, error) {
 			)
 		}
 
+		if flags.Redis != "" {
```
Comment (jenlij) [inline cmd/storj-sim/network.go]: I'm not sure if this is correct

File: pkg/revocation/db_test.go [hunk 7]
```diff
@@ -151,3 +151,57 @@ func TestRevocationDB_Put_error(t *testing.T) {
...
+		secondRevocation, err := extensions.NewRevocationExt(keys2[peertls.CAIndex], chain2[peertls.LeafIndex])
+		require.NoError(t, err)
+		err = revDB.Put(ctx, chain2, secondRevocation)
+		require.NoError(t, err)
+		revs, err = revDB.List(ctx)
+		assert.NoError(t, err)
+		assert.Equal(t, 2, len(revs))
+
+		expected := [][]byte{firstRevocation.Value, secondRevocation.Value}
```
Comment (bryanchriswhite) [inline pkg/revocation/db_test.go]: Nice solution to the order problem! Would `assert.Contains` work for the assertion instead of having to roll your own?

File: pkg/revocation/common.go [hunk 8]
```diff
@@ -27,23 +29,23 @@ func NewDB(dbURL string) (*DB, error) {
...
 	}
-
 	var db *DB
 	switch driver {
 	case "bolt":
 		db, err = newDBBolt(source)
 		if err != nil {
 			return nil, extensions.ErrRevocationDB.Wrap(err)
 		}
+		fmt.Println("JEN : bolt")
```
Comment (jenlij) [inline pkg/revocation/common.go]: remember to remove these print statements

File: cmd/storj-sim/network.go (hunk 6 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: storj-sim network --redis="redis://127.0.0.1:6379?db=0" setup fails to create a config file

File: pkg/revocation/db_test.go (hunk 7 above)
Comment (jenlij) [reply pkg/revocation/db_test.go]: yep that does work, thanks!

**Comment by jenlij** [comment]
Note I need to start a redis server for integration tests

**Comment by bryanchriswhite** [comment]
> Note I need to start a redis server for integration tests

I think you can use this: https://github.com/storj/storj/blob/856b98997c580d71da4b636ff67679aad1038a8d/storage/redis/redisserver/server.go#L48

File: cmd/storj-sim/network.go [hunk 9]
```diff
@@ -250,7 +250,7 @@ func newNetwork(flags *Flags) (*Processes, error) {
 				"--server.address", process.Address,
 				"--server.private-address", net.JoinHostPort(host, port(satellitePeer, i, privateGRPC)),
 
-				"--server.extensions.revocation=false",
+				"--server.extensions.revocation=false", //todo update to true
```
Comment (JessicaGreben) [inline cmd/storj-sim/network.go]: should this say "update to true when we want to use revocation" or something?

Comment (jenlij) [reply cmd/storj-sim/network.go]: yeah I could add that. I'm wondering if we want to keep revocationdb once we have a redis namespace package so we can make sure it works with the satellite separation?

File: cmd/storj-sim/network.go [hunk 10]
```diff
@@ -269,6 +269,11 @@ func newNetwork(flags *Flags) (*Processes, error) {
 			)
 		}
 
+		if flags.Redis != "" {
+			process.Arguments["setup"] = append(process.Arguments["setup"],
+				"--server.revocation-dburl", flags.Redis)
```
Comment (bryanchriswhite) [inline cmd/storj-sim/network.go]: If `flags.Redis` is supposed to be used "for all databases that support it" we probably need to use different dbs (`?db=0`, `?db=1`, etc.) for the different databases, otherwise I think all keys will share a namespace which could be problematic.

File: cmd/storj-sim/network.go (hunk 9 above)
Comment (bryanchriswhite) [reply cmd/storj-sim/network.go]: Just fyi, this config flag only determines [whether we check for revocation extensions on peer certificates during TLS handshakes](https://github.com/storj/storj/blob/master/pkg/peertls/tlsopts/options.go#L107), which is different from whether we're opening or [passing a revocation DB to the server run method](https://github.com/storj/storj/blob/94cec7e3b05b06b522697e206d6e29b6b45156ee/pkg/server/config.go#L26).

Also, just curious what you mean by "wondering if we want to keep revocationdb..."?

Comment (jenlij) [reply cmd/storj-sim/network.go]: oh ok good to know. and yeah that makes little sense now that I re-read that haha... I meant if we wanted to keep that flag=true.

File: cmd/storj-sim/network.go (hunk 10 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: Yeah definitely, I am opening another PR to be able to designate different namespaces for difference processes. So we can have revocationdb=db0, liveaccounting=db1...

Comment (jenlij) [reply cmd/storj-sim/network.go]: I can try to get that in before this pr if that's more strategic

Comment (bryanchriswhite) [reply cmd/storj-sim/network.go]: Maybe just leave a todo for now then?

Comment (jenlij) [reply cmd/storj-sim/network.go]: sounds good!

File: pkg/revocation/db.go [hunk 11]
```diff
@@ -112,7 +112,9 @@ func (db *DB) List(ctx context.Context) (revs []*extensions.Revocation, err erro
 	if err != nil {
 		return nil, extensions.ErrRevocationDB.Wrap(err)
 	}
-
+	if len(keys) == 0 {
```
Comment (egonelbre) [inline pkg/revocation/db.go]: Why is this case here?

All of the following should work when there aren't any keys. If it doesn't then it's probably a bug in the store implementation rather than a mistake here.

Comment (bryanchriswhite) [reply pkg/revocation/db.go]: As far as I understood, the redis client returns an error in this case.

Comment (jenlij) [reply pkg/revocation/db.go]: mget errors when you pass in an empty slice https://github.com/NodeRedis/node_redis/issues/125

Comment (egonelbre) [reply pkg/revocation/db.go]: Generally yes we should try to homogenize the different implementation rather than push the concern to the user of the interface. The main reason is that otherwise the abstraction gets leaky, and we would end up having these "exceptional cases" for every caller and every implementation.

File: cmd/storj-sim/network.go (hunk 10 above)
Comment (egonelbre) [inline cmd/storj-sim/network.go]: One important thing here is that we can have multiple satellites and they should not share the same database.

File: pkg/revocation/db.go (hunk 11 above)
Comment (jenlij) [reply pkg/revocation/db.go]: I'll leave a note then that this happens

File: cmd/storj-sim/network.go (hunk 10 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: oh good point

File: pkg/revocation/db.go (hunk 11 above)
Comment (egonelbre) [reply pkg/revocation/db.go]: I usually create a separate PR that fixes rather than leaving a note. Otherwise we'll end up collecting so many of those todo notes :D

File: cmd/storj-sim/network.go (hunk 10 above)
Comment (bryanchriswhite) [reply cmd/storj-sim/network.go]: We might want them to share a revocation db though.

Comment (egonelbre) [reply cmd/storj-sim/network.go]: Not really, the purpose here is more to ensure that things work with separate satellites that have no knowledge of each other.

Comment (bryanchriswhite) [reply cmd/storj-sim/network.go]: My point is that it would be beneficial to share knowledge of revisions across satellites.

Comment (egonelbre) [reply cmd/storj-sim/network.go]: How do we do that when we have ten partners each with their own revocation database implementation running independently none of which have direct access to each other?

Comment (bryanchriswhite) [reply cmd/storj-sim/network.go]: I'm only taking about stuff we control. Obviously we can't do that in your example.

Comment (egonelbre) [reply cmd/storj-sim/network.go]: Yeah, I agree. The purpose here simulating multiple satellites is to simulate independent satellites. Sure we could also try to simulate cooperating satellites, but that in my mind would be secondary.

File: cmd/storj-sim/main.go [hunk 12]
```diff
@@ -59,6 +62,9 @@ func main() {
 
 	rootCmd.PersistentFlags().StringVarP(&flags.Postgres, "postgres", "", os.Getenv("STORJ_SIM_POSTGRES"), "connection string for postgres (defaults to STORJ_SIM_POSTGRES)")
 
+	// todo make redis required
+	rootCmd.PersistentFlags().StringVarP(&flags.Redis, "redis", "", "", "connection string for redis. If provided, storj-sim will use redis for all databases that support it.")
```
Comment (thepaul) [inline cmd/storj-sim/main.go]: Is "all databases that support it" meant to include the metainfoDB (pointerDB)? I think that would be supported by our redis client, but I don't think this PR makes that happen (also, I'm not sure that I would want that to happen).

File: pkg/revocation/db_test.go [hunk 13]
```diff
@@ -151,3 +151,48 @@ func TestRevocationDB_Put_error(t *testing.T) {
...
+
+	testrevocation.RunDBs(t, func(t *testing.T, revDB extensions.RevocationDB, db storage.KeyValueStore) {
+		keys, chain, err := testpeertls.NewCertChain(2, storj.LatestIDVersion().Number)
+		require.NoError(t, err)
+		keys2, chain2, err := testpeertls.NewCertChain(2, storj.LatestIDVersion().Number)
+		require.NoError(t, err)
+
+		// test list no revocations, should not error
+		revs, err := revDB.List(ctx)
+		assert.NoError(t, err)
```
Comment (thepaul) [inline pkg/revocation/db_test.go]: `assert.NoError()` is almost never appropriate. If there is an error here, the things that happen afterward are not expected to work at all. Use `require.NoError()` here (and for all other `assert.NoError()` instances in this file) so that the test exits on error instead of trying to continue.

File: pkg/revocation/db_test.go [hunk 14]
```diff
@@ -151,3 +151,48 @@ func TestRevocationDB_Put_error(t *testing.T) {
...
+
+		// list 1,2 revocations
+		firstRevocation, err := extensions.NewRevocationExt(keys[peertls.CAIndex], chain[peertls.LeafIndex])
+		require.NoError(t, err)
+
+		err = revDB.Put(ctx, chain, firstRevocation)
+		require.NoError(t, err)
+		revs, err = revDB.List(ctx)
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(revs))
```
Comment (thepaul) [inline pkg/revocation/db_test.go]: Might want to `require` here as well. If `len(revs)` is actually 0, and execution continues, you'll hit a panic on the next line (which will hide the assertion error that was added on this line, causing confusion).

It might even be a good idea to use `require` exclusively in this file. The `assert` package is only helpful when you are sure nothing will crash if execution carries on after an assertion failure, and when there is reason to think it's worthwhile for execution to carry on.

File: pkg/revocation/db_test.go (hunk 13 above)
Comment (jenlij) [reply pkg/revocation/db_test.go]: good to know!

File: cmd/storj-sim/main.go (hunk 12 above)
Comment (jenlij) [reply cmd/storj-sim/main.go]: hmm yeah I agree that's not a great description. will update

File: cmd/storj-sim/network.go (hunk 10 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: i'll see if i can spin up a unique miniredis server for each satellite

File: storage/redis/redisnamespace/redisnamespace.go [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+// See LICENSE for copying information.
+
+package redisnamespace
+
+import "strconv"
+
+var dbs = map[string]int{
+	"live-accounting.db":      0,
+	"server.revocation-dburl": 1,
+}
```
Comment (JessicaGreben) [inline storage/redis/redisnamespace/redisnamespace.go]: why a map here instead of `const`?

Comment (jenlij) [reply storage/redis/redisnamespace/redisnamespace.go]: I wanted to be able to iterate over them and you can't (as far as I know) do that with `const`.

File: cmd/storj-sim/network.go [hunk 16]
```diff
@@ -177,10 +189,10 @@ func networkDestroy(flags *Flags, args []string) error {
 }
 
 // newNetwork creates a default network
-func newNetwork(flags *Flags) (*Processes, error) {
+func newNetwork(flags *Flags) (processes *Processes, cleanup func(), err error) {
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: Adding this `cleanup` doesn't look that nice.

File: cmd/storj-sim/network.go [hunk 17]
```diff
@@ -217,12 +229,21 @@ func newNetwork(flags *Flags) (*Processes, error) {
 	versioncontrol.ExecBefore["run"] = func(process *Process) error {
 		return readConfigString(&versioncontrol.Address, versioncontrol.Directory, "address")
 	}
-
 	// gateway must wait for the versioncontrol to start up
 
+	// start a test redis server
+	rootpath := flags.Redis
+	if flags.Redis == "testServer" {
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: why not instead check for empty string?

File: cmd/storj-sim/network.go [hunk 18]
```diff
@@ -217,12 +229,21 @@ func newNetwork(flags *Flags) (*Processes, error) {
...
+		address, cleanup, err := redisserver.Start()
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: each satellite should probably have a separate redis instance

Comment (jenlij) [reply cmd/storj-sim/network.go]: ok

File: cmd/storj-sim/network.go (hunk 17 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: empty string could also indicate we want to use bolt... unless we want to eliminate that option

File: cmd/storj-sim/network.go (hunk 16 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: where can I call cleanup instead?

File: cmd/storj-sim/network.go (hunk 17 above)
Comment (egonelbre) [reply cmd/storj-sim/network.go]: I think we can stop worrying about bolt.

File: satellite/accounting/live/cache.go [hunk 19]
```diff
@@ -21,13 +21,13 @@ var (
 
 // Config contains configurable values for the live accounting service.
 type Config struct {
-	StorageBackend string `help:"what to use for storing real-time accounting data" default:"memory"`
+	DB string `help:"what to use for storing real-time accounting data" default:"memory"`
```
Comment (JessicaGreben) [inline satellite/accounting/live/cache.go]: do we want to change this config name? we will need to change it for all the production satellites, etc

File: Jenkinsfile.public [hunk 20]
```diff
@@ -26,7 +26,7 @@ pipeline {
                 sh 'go mod download'
```
Comment (jenlij) [inline Jenkinsfile.public]: we won't need this file if we remove the redis flag from storj sim

File: scripts/test-sim.sh [hunk 21]
```diff
@@ -18,13 +18,17 @@ export STORJ_NETWORK_DIR=$TMP
 
 STORJ_NETWORK_HOST4=${STORJ_NETWORK_HOST4:-127.0.0.1}
 STORJ_SIM_POSTGRES=${STORJ_SIM_POSTGRES:-""}
+STORJ_SIM_REDIS=${STORJ_SIM_REDIS:-""}
```
Comment (jenlij) [inline scripts/test-sim.sh]: we won't need these file changes if we remove the redis flag from storj sim

File: storage/redis/client.go [hunk 22]
```diff
@@ -137,7 +139,7 @@ func (client *Client) GetAll(ctx context.Context, keys storage.Keys) (_ storage.
 
 	results, err := client.db.MGet(keyStrings...).Result()
 	if err != nil {
-		return nil, err
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (jenlij) [reply storage/redis/client.go]: To address this issue https://github.com/storj/storj/pull/3259/files?file-filters%5B%5D=.go&file-filters%5B%5D=.public#discussion_r337673160

File: cmd/storj-sim/network.go [hunk 23]
```diff
@@ -217,14 +220,51 @@ func newNetwork(flags *Flags) (*Processes, error) {
...
+		redisPort = port(satellitePeer, i, redisEndpoint)
+		process := processes.New(Info{
+			Name:       fmt.Sprintf("redis/%d", i),
+			Executable: "redis-server",
+			Directory:  filepath.Join(processes.Directory, "satellite", fmt.Sprint(i), "redis"),
+			Address:    net.JoinHostPort(host, redisPort),
+		})
+		redisServers = append(redisServers, process)
+
+		filename := process.Directory + "sim.rdb"
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: When putting paths together use `filepath`

File: cmd/storj-sim/network.go [hunk 24]
```diff
@@ -217,14 +220,51 @@ func newNetwork(flags *Flags) (*Processes, error) {
...
+				"bind" + host,
+				"port " + redisPort,
+				"timeout 0",
+				"databases" + strconv.Itoa(len(redisDBs)),
+				"dbfilename" + filename,
+				"dir " + process.Directory,
+			}
+			conf := strings.Join(arguments, "\n") + "\n"
+			err = ioutil.WriteFile(confpath, []byte(conf), 0755)
+			if err != nil {
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: just `return err`

File: cmd/storj-sim/network.go [hunk 25]
```diff
@@ -177,7 +180,7 @@ func networkDestroy(flags *Flags, args []string) error {
 }
 
 // newNetwork creates a default network
-func newNetwork(flags *Flags) (*Processes, error) {
+func newNetwork(flags *Flags) (processes *Processes, err error) {
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: Why change this?

File: storage/redis/redisnamespace/redisnamespace.go [hunk 26]
```diff
@@ -0,0 +1,21 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package redisnamespace
+
+import "strconv"
+
+var dbs = map[string]int{
```
Comment (egonelbre) [inline storage/redis/redisnamespace/redisnamespace.go]: Why define these here when only storj-sim needs them?

File: cmd/storj-sim/network.go (hunk 25 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: that was remanent of when I tried returning "cleanup" I have yet to push the update

File: storage/redis/redisnamespace/redisnamespace.go (hunk 26 above)
Comment (jenlij) [reply storage/redis/redisnamespace/redisnamespace.go]: good point, I started them for general use, but yeah only storj-sim is using it atm

File: cmd/storj-sim/network.go [hunk 27]
```diff
@@ -268,10 +308,12 @@ func newNetwork(flags *Flags) (*Processes, error) {
 				"--metainfo.database-url", pgutil.ConnstrWithSchema(flags.Postgres, fmt.Sprintf("satellite/%d/meta", i)),
 			)
 		}
-
-		process.ExecBefore["run"] = func(process *Process) error {
-			return readConfigString(&process.Address, process.Directory, "server.address")
+		for name, db := range redisDBs {
+			flag := "--" + name
+			url := redisnamespace.CreatePath(net.JoinHostPort(host, redisPort), db)
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: You can hardcode the paths instead

File: storage/redis/client.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply storage/redis/client.go]: The solution here is to check:
```
if len(keyStrings) == 0 {
    return nil, nil
}
```

Rather than ignore the error itself.

[the remark this answers is not in the document]
Comment (egonelbre) [reply storage/redis/client.go]: Move this before the query to the database, since it probably doesn't make sense to make do the query in this case.

File: cmd/storj-sim/network.go [hunk 28]
```diff
@@ -61,6 +62,7 @@ const (
 	// satellite specific constants
 	debugPeerHTTP     = 7
 	debugRepairerHTTP = 8
+	redisConst        = 4
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: You can call this `redisPort`

File: cmd/storj-sim/network.go [hunk 29]
```diff
@@ -222,14 +225,47 @@ func newNetwork(flags *Flags) (*Processes, error) {
...
 	// gateway must wait for the versioncontrol to start up
 
 	// Create satellites
 	if flags.SatelliteCount > maxInstanceCount {
 		return nil, fmt.Errorf("exceeded the max instance count of %d with Satellite count of %d", maxInstanceCount, flags.SatelliteCount)
 	}
 
+	// set up redis servers
+	var redisServers []*Process
+	var redisPort string
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: Hmm, why is this here? There are multiple redis server and each of them has a separate port.

Comment (jenlij) [reply cmd/storj-sim/network.go]: oh for the actual port value like 10014

File: cmd/storj-sim/network.go [hunk 30]
```diff
@@ -293,7 +330,7 @@ func newNetwork(flags *Flags) (*Processes, error) {
 				"--debug.addr", net.JoinHostPort(host, port(satellitePeer, i, debugPeerHTTP)),
 			},
 		})
-
+		process.WaitForStart(redisServers[i])
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: Since it's already waiting for the satellite to start-up, there probably isn't a need to wait for the redis-server to start separately.

Comment (jenlij) [reply cmd/storj-sim/network.go]: yep i removed it... i think i may have pushed right as you commented, so there are recent changes

File: cmd/storj-sim/network.go (hunk 29 above)
Comment (egonelbre) [reply cmd/storj-sim/network.go]: Yes, but there will be multiple of them. You can rely on `port(satellitePeer, redisTCP)` for the time being everywhere. Or `redisServers[i].Address`

File: cmd/storj-sim/network.go [hunk 31]
```diff
@@ -310,7 +311,7 @@ func newNetwork(flags *Flags) (*Processes, error) {
 			)
 		}
 		for flag, db := range redisDBs {
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: note, you can inline these

File: cmd/storj-sim/network.go (hunk 29 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: hmm I'm not sure I follow, the way I have it right now each satellite will use a separate redis port numbered 100X4, which was generated by the `port` method. I keep track of them using `ports[]`

Comment (egonelbre) [reply cmd/storj-sim/network.go]: `port` is idempotent and so is `redisServers[i].Address` that way you don't need to have a separate `[]string` to track them.

Comment (jenlij) [reply cmd/storj-sim/network.go]: oh I see, the github history is misleading, I thought you were implying all satellites could use the same redis server/port.

File: cmd/storj-sim/network.go (hunk 31 above)
Comment (jenlij) [reply cmd/storj-sim/network.go]: ok, in that case, should i just change the redisDBs map to just be a `const`?

File: cmd/storj-sim/network.go (hunk 29 above)
Comment (egonelbre) [reply cmd/storj-sim/network.go]: I still haven't had my morning coffee to properly explain things :D

Comment (jenlij) [reply cmd/storj-sim/network.go]: :D no worries at all, it's past midnight here so I'm misreading things lol

File: cmd/storj-sim/network.go (hunk 31 above)
Comment (egonelbre) [reply cmd/storj-sim/network.go]: To the arguments starting at lines 282, you can add:

```
"--live-accounting.storage-backend": "redis://" + redisServers[i] + "?db=0",
"--server.revocation-dburl":         "redis://" + redisServers[i] + "?db=1",
```

Comment (jenlij) [reply cmd/storj-sim/network.go]: perfect

File: cmd/storj-sim/network.go [hunk 32]
```diff
@@ -273,10 +310,7 @@ func newNetwork(flags *Flags) (*Processes, error) {
 				"--metainfo.database-url", pgutil.ConnstrWithSchema(flags.Postgres, fmt.Sprintf("satellite/%d/meta", i)),
 			)
 		}
-
-		process.ExecBefore["run"] = func(process *Process) error {
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: This needs to stay in place
